haystack/core/pipeline/pipeline.py

```diff
@@ -349,204 +185,81 @@ def run(  # noqa: PLR0915, PLR0912
         """
         pipeline_running(self)
 
-        # Reset the visits count for each component
-        self._init_graph()
-
         # TODO: Remove this warmup once we can check reliably whether a component has been warmed up or not
```

Comment:
As far as I've understood, this is to avoid a component being "warmed up" several times, which can happen if a pipeline is run multiple times - If it's not already handled, maybe it's worth opening an issue for this.

Reply:
No, this only reset the component visits to 0. The warm_up is handled in the `warm_up`-method (https://github.com/deepset-ai/haystack/blob/82d95e0bc75f6f0c623a43c36957c37f002fc4d2/haystack/core/pipeline/base.py#L725).

We can remove resetting the visits to 0 because we are not tracking visits as instance state anymore. Instead, we track visits on a per-run invocation basis.

Previously, pipeline.run was not thread-safe because the visits were tracked on the instance level, so one run invocation would impact the other.

Reply:
sorry, my comment is related to the `self.warm_up()` below - I just started writing it in association with the `# TODO` line - we should check ( open an issue to work on this) if a component was warmed up already, before doing so, since this can delay a pipeline run.

Reply:
It's not a blocker - I just wanted to have a reminder/issue so that we can later work on this.

We can agree on how to handle those. For instance, every component that has a `warm_up()` method should also have a `self.warmed_up : bool = True|False`

Reply:
@mathislucka I will open an issue to track this

Reply:
https://github.com/deepset-ai/haystack/issues/8823